src/backend/d3d12/PipelineLayoutD3D12.cpp

```diff
@@ -24,10 +25,59 @@ namespace d3d12 {
     PipelineLayout::PipelineLayout(Device* device, PipelineLayoutBuilder* builder)
         : PipelineLayoutBase(builder), device(device) {
 
-        // Create an empty root signature.
+        D3D12_ROOT_PARAMETER rootParameters[kMaxBindGroups * 2];
+
+        // A root parameter is one of these types
+        union {
+            D3D12_ROOT_DESCRIPTOR_TABLE DescriptorTable;
+            D3D12_ROOT_CONSTANTS Constants;
+            D3D12_ROOT_DESCRIPTOR Descriptor;
+        } rootParameterValues[kMaxBindGroups * 2];
+        // samplers must be in a separate descriptor table so we need at most twice as many tables as bind groups
+
+        // Ranges are D3D12_DESCRIPTOR_RANGE_TYPE_(SRV|UAV|CBV|SAMPLER)
+        // They are grouped together so each bind group has at most 4 ranges
+        D3D12_DESCRIPTOR_RANGE ranges[kMaxBindGroups * 4];
+
+        uint32_t parameterIndex = 0;
+        uint32_t rangeIndex = 0;
+
+        for (uint32_t group = 0; group < kMaxBindGroups; ++group) {
+            const BindGroupLayout* bindGroupLayout = ToBackend(GetBindGroupLayout(group));
+
+            // Set the root descriptor table parameter and copy ranges. Ranges are offset by the bind group index
+            // Returns whether or not the parameter was set. A root parameter is not set if the number of ranges is 0
+            auto SetRootDescriptorTable = [&](uint32_t rangeCount, const D3D12_DESCRIPTOR_RANGE* descriptorRanges) -> bool {
+                if (rangeCount > 0) {
```

Comment:
This could be simplified by having an early return:

```cpp
if (rangeCount == 0) {
    return false;
}
// set root parameter
// Loop over ranges
 return true;
```